proof-systems/src/darlin/marlin/src/lib.rs

```diff
@@ -124,16 +126,29 @@ impl<G: Group, PC: PolynomialCommitment<G>> Marlin<G, PC> {
             .map(|c| c.commitment().clone())
             .collect();
 
-        let vk_hash = D::digest(&serialize_no_metadata![index.index_info, index_comms]
-            .map_err(|e| {
-                Error::Other(format!("Unable to serialize vk elements: {:?}", e))
-            })?)
-            .as_ref()
-            .to_vec();
+        let vk_hash = D::digest(
+            &serialize_no_metadata![index.index_info, index_comms]
+                .map_err(|e| Error::Other(format!("Unable to serialize vk elements: {:?}", e)))?,
+        )
+        .as_ref()
+        .to_vec();
+
+        // Compute the commitments of the Lagrange polynomials over the input domain.
+        // They are included into the verifier key in order to help the verifier check that the
+        // polynomial behind the commitment of the input poly provided by the prover is indeed the
+        // input poly.
+        let domain_x = get_best_evaluation_domain(index.index_info.num_inputs).unwrap();
+        let lagrange_comms: Vec<_> = domain_x
+            .elements()
+            .into_iter()
+            .map(|y| domain_x.slice_lagrange_kernel(y))
+            .map(|poly| PC::commit(committer_key, &poly, false, None).unwrap().0)
```

Comment:
- Merged the two maps together
- The way `algebra::fft::domain::Elements` is currently implemented does not easily allow to parallelize the iteration. It would be nice to make this possible, but not really needed for this specific use case, since public input should always be relatively small.